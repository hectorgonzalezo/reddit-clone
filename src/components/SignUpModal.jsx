import React, { useState, useRef } from 'react';
import Button from './Button';
import googleIcon from '../assets/google.png';
import '../styles/signUpModal.scss';
import { authorization } from '../firebase/firebase';

function SignUpModal() {
  const formRef = useRef();
  const password1Ref = useRef();
  const password2Ref = useRef();
  const [disableButton, setDisableButton] = useState(true);

  // This function gets called on every input value change
  // If the whole form is valid, it activates the continue button
  function validate(e) {
    // Check input validity
    const elementValidity = e.target.validity;
    // Checkt type of validity
    switch (true) {
      case elementValidity.typeMismatch:
        e.target.setCustomValidity('Please write a valid email');
        e.target.parentNode.firstChild.innerText = 'Please write a valid email';
        break;
      case elementValidity.patternMismatch:
        e.target.setCustomValidity('Only letters allowed in username');
        e.target.parentNode.firstChild.innerText = 'Only letters allowed in username';
        break;
      case elementValidity.tooShort:
        e.target.setCustomValidity('Password must be at least 6 characters long');
        e.target.parentNode.firstChild.innerText = 'Password must be at least 6 characters long';
        break;
      default:
        e.target.setCustomValidity('');
        e.target.parentNode.firstChild.innerText = '';
    }

    // console.log(elementValidity);

    // if element is password, check that both passwords equal each other
    if ((e.target === password2Ref.current || e.target === password1Ref.current)) {
      if (
        // writing second password and first one is not the same
        (e.target === password2Ref.current
        && password2Ref.current.value !== password1Ref.current.value)
        // writing first password and second one is not empty
        || (e.target === password1Ref.current
          && password2Ref.current.value !== ''
          && password2Ref.current.value !== password1Ref.current.value)
      ) {
        password2Ref.current.setCustomValidity("Passwords don't match");
        password2Ref.current.parentNode.firstChild.innerText = "Passwords don't match";
      } else {
        password2Ref.current.setCustomValidity('');
        password2Ref.current.parentNode.firstChild.innerText = '';
      }
    }

    // if form is valid, activate submit button
    setDisableButton(!formRef.current.checkValidity());
  }

  function submitSignUp(e) {
    e.preventDefault();
    console.log('signup')
  }

  return (
    <div id="signUp-outer">
      <div id="signUp-inner">
        <div>
          <button type="button" className="close-button">
            x
          </button>
          <h1>Sign Up</h1>
          <p>
            By continuing, you are setting up a Reddit account and agree to our{" "}
            <a href="https://www.redditinc.com/policies/user-agreement">
              User Agreement
            </a>{" "}
            and{" "}
            <a href="https://www.reddit.com/policies/privacy-policy">
              Privacy Policy
            </a>
            .
          </p>
          <button onClick={authorization.logInPopup} className="google-button">
            <img src={googleIcon} alt="" />
            Continue with Google
          </button>
        </div>

        <div className="division">
          <hr />
          <h3>or</h3>
          <hr />
        </div>
        <form action="" ref={formRef}>
          <div className="input-wrap">
            <span />
            <input
              type="text"
              placeholder="0"
              name="username"
              id="username"
              onChange={validate}
              pattern="[A-Za-z0-9]*"
              maxLength={25}
              required
            />
            <label htmlFor="username">Username</label>
          </div>
          <div className="input-wrap">
            <span />
            <input
              type="email"
              placeholder="0"
              name="email"
              id="email"
              onChange={validate}
              required
            />
            <label htmlFor="email">Email</label>
          </div>
          <div className="input-wrap">
            <span />
            <input
              type="password"
              placeholder="0"
              name="password"
              id="password"
              ref={password1Ref}
              onChange={validate}
              minLength={6}
              required
            />
            <label htmlFor="password">Password</label>
          </div>
          <div className="input-wrap">
            <span />
            <input
              type="password"
              placeholder="0"
              name="repeatPassword"
              id="repeatPassword"
              ref={password2Ref}
              onChange={validate}
              minLength={6}
              required
            />
            <label htmlFor="repeatPassword">Repeat Password</label>
          </div>
          <Button text="Sign Up" type="submit" disabled={disableButton} onClick={submitSignUp} />
        </form>
        <p>
          Already a redditor? <a href="">Sign In</a>
        </p>
      </div>
    </div>
  );
}

export default SignUpModal;
