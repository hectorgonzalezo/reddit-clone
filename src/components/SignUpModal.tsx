import React, { useState, useRef, SyntheticEvent } from 'react';
import { signUp } from '../api/users';
import { useDispatch } from 'react-redux';
import Button from './Button';
import { addUser } from '../store/userSlice';
import loadingIcon from '../assets/loading.gif';

interface SignUpModalProps {
  closeFunc: () => void;
};

function SignUpModal({ closeFunc= () => {} }: SignUpModalProps): JSX.Element {
  const formRef = useRef<HTMLFormElement>(null);
  const userNameRef = useRef<HTMLInputElement>(null);
  const emailRef = useRef<HTMLInputElement>(null);
  const password1Ref = useRef<HTMLInputElement>(null);
  const password2Ref = useRef<HTMLInputElement>(null);
  const [disableButton, setDisableButton] = useState(true);
  const [emailAlreadyExists, setEmailAlreadyExists] = useState(false);
  const [loadingData, setLoadingData] = useState(false);
  const dispatch = useDispatch();

  // This function gets called on every input value change
  // If the whole form is valid, it activates the continue button
  function validate(e: SyntheticEvent): void {
    const target = e.target as HTMLInputElement;
    const form = formRef.current as HTMLFormElement;
    // Check input validity
    const elementValidity = target.validity;
    if (target.parentNode?.firstChild) {
      // error display element
      const display = target.parentNode.firstChild as HTMLSpanElement;
    // Check type of validity
    switch (true) {
      case elementValidity.typeMismatch:
        target.setCustomValidity('Please write a valid email');
        display.innerText = 'Please write a valid email';
        break;
      case elementValidity.patternMismatch:
        if (target === userNameRef.current) {
          target.setCustomValidity('Only letters allowed in username');
          display.innerText = 'Only letters allowed in username';
        } else {
          target.setCustomValidity('Please write a valid email');
          display.innerText = 'Please write a valid email';
        }
        break;
      case elementValidity.tooShort:
        target.setCustomValidity('Password must be at least 6 characters long');
        display.innerText = 'Password must be at least 6 characters long';
        break;
      default:
        target.setCustomValidity('');
        display.innerText = '';
    }
  }


    // if element is password, check that both passwords equal each other
    if (
      (target === password2Ref.current || target === password1Ref.current) &&
      password1Ref?.current !== null &&
      password2Ref?.current?.parentNode
    ) {
      const display = password2Ref.current.parentNode.firstChild as HTMLSpanElement;
      if (
        // writing second password and first one is not the same
        (target === password2Ref.current &&
          password2Ref.current.value !== password1Ref.current.value) ||
        // writing first password and second one is not empty
        (target === password1Ref.current &&
          password2Ref.current.value !== "" &&
          password2Ref.current.value !== password1Ref.current.value)
      ) {
        password2Ref.current.setCustomValidity("Passwords don't match");
        display.innerText =
          "Passwords don't match";
      } else {
        password2Ref.current.setCustomValidity("");
        display.innerText = "";
      }
    }

    // if form is valid, activate submit button
    setDisableButton(!form.checkValidity());
  }

  function submitSignUp(e: SyntheticEvent): void {
    e.preventDefault();
    if (
      userNameRef?.current !== null &&
      emailRef.current !== null &&
      password1Ref.current !== null &&
      password2Ref.current !== null
    ) {
      const username = userNameRef.current.value;
      const email = emailRef.current.value;
      const password = password1Ref.current.value;
      const passwordConfirm = password2Ref.current.value;
      const icon =
        "https://firebasestorage.googleapis.com/v0/b/reddit-clone-83ce9.appspot.com/o/user_icon.svg?alt=media&token=50e7a9f1-8508-4d51-aac8-4d1ed9dad7a1";
      const votes = {};
      try {
        // add loading animation
        setLoadingData(true);

        // Sign up user
        signUp({ username, email, password, passwordConfirm })
          .then((data) => {
            setEmailAlreadyExists(false);
            // update redux store
            dispatch(addUser(data));
            closeFunc();
          })
          .catch((err) => console.log(err));
      } catch (error) {
        // If email already exists
        setEmailAlreadyExists(true);
        setLoadingData(false);
      }
    }
  }

  return (
    <div id="signUp-outer" className="modal-outer">
      <div id="signUp-inner" className="modal-inner">
        <div>
          <button type="button" className="close-button" onClick={closeFunc} aria-label="close button">
            x
          </button>
          <h1>Sign Up</h1>
          <p>
            By continuing, you are setting up a Reddit account and agree to our{" "}
            <a href="https://www.redditinc.com/policies/user-agreement">
              User Agreement
            </a>{" "}
            and{" "}
            <a href="https://www.reddit.com/policies/privacy-policy">
              Privacy Policy
            </a>
            .
          </p>
        </div>

        <div className="division"/>
        <form action="" ref={formRef}>
          <div className="input-wrap">
            <span />
            <input
              type="text"
              placeholder="0"
              name="username"
              id="username"
              ref={userNameRef}
              onChange={validate}
              pattern="[A-Za-z0-9]*"
              maxLength={25}
              required
            />
            <label htmlFor="username">Username</label>
          </div>
          <div className="input-wrap">
            <span />
            <input
              type="email"
              placeholder="0"
              name="email"
              ref={emailRef}
              pattern="[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}$"
              id="email"
              onChange={validate}
              required
            />
            <label htmlFor="email">Email</label>
          </div>
          <div className="input-wrap">
            <span />
            <input
              type="password"
              placeholder="0"
              name="password"
              id="password"
              ref={password1Ref}
              onChange={validate}
              minLength={6}
              autoComplete="on"
              required
            />
            <label htmlFor="password">Password</label>
          </div>
          <div className="input-wrap">
            <span />
            <input
              type="password"
              placeholder="0"
              name="repeatPassword"
              id="repeatPassword"
              ref={password2Ref}
              onChange={validate}
              minLength={6}
              autoComplete="on"
              required
            />
            <label htmlFor="repeatPassword">Repeat Password</label>
          </div>
          {emailAlreadyExists ? <span>Email is already registered</span> : null}
          <Button
            text=""
            type="submit"
            disabled={disableButton}
            onClick={submitSignUp}
          >
            {loadingData ? (
              <img src={loadingIcon} alt="loading" data-testid="loading-icon" className="loading-icon" />
            ) : (
              ["Sign Up"]
            )}
          </Button>
        </form>
        <p>
          Already a redditor? <a href="">Sign In</a>
        </p>
      </div>
    </div>
  );
}



export default SignUpModal;
