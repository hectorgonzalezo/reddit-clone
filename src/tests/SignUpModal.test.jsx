
import React from 'react';
import { render, screen, container } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import SignUpModal from '../components/SignUpModal';

describe('Styled button works', () => {
  test('Renders with sign up button disabled', () => {
    render(<SignUpModal />);
    const button = screen.getByRole('button', { name: 'Sign Up' });
    expect(button).toHaveStyle('opacity: 0.5');
    expect(button).toHaveProperty('disabled', true);

    expect(container).toMatchSnapshot();
  });

  test('After filling form correctly, button gets enabled', () => {
    render(<SignUpModal />);
    const textInputs = screen.queryAllByRole('textbox');
    const passwordInputs = screen.queryAllByLabelText(/password/i);

    // Fill in form
    userEvent.type(textInputs[0], 'mockUser');
    userEvent.type(textInputs[1], 'mock@mock.com');
    passwordInputs.forEach((input) => userEvent.type(input, 'password'));

    // Button should be enabled
    const button = screen.getByRole('button', { name: 'Sign Up' });
    expect(button).not.toHaveStyle('opacity: 0.5');
    expect(button).toHaveProperty('disabled', false);

    expect(container).toMatchSnapshot();
  });

  test('Button doesnt get enabled if mail is missing @', () => {
    render(<SignUpModal />);
    const textInputs = screen.queryAllByRole('textbox');
    const passwordInputs = screen.queryAllByLabelText(/password/i);

    // Fill in form
    userEvent.type(textInputs[0], 'mockUser');
    userEvent.type(textInputs[1], 'mock');
    passwordInputs.forEach((input) => userEvent.type(input, 'password'));

    // Button should be enabled
    const button = screen.getByRole('button', { name: 'Sign Up' });
    expect(button).toHaveStyle('opacity: 0.5');
    expect(button).toHaveProperty('disabled', true);

    expect(container).toMatchSnapshot();
  });

  test('Button doesnt get enabled if passwords are different', () => {
    render(<SignUpModal />);
    const textInputs = screen.queryAllByRole('textbox');
    const passwordInputs = screen.queryAllByLabelText(/password/i);

    // Fill in form
    userEvent.type(textInputs[0], 'mockUser');
    userEvent.type(textInputs[1], 'mock@mock.com');
    passwordInputs.forEach((input) => userEvent.type(input, `password${Math.random()}`));

    // Button should be enabled
    const button = screen.getByRole('button', { name: 'Sign Up' });
    expect(button).toHaveStyle('opacity: 0.5');
    expect(button).toHaveProperty('disabled', true);

    expect(container).toMatchSnapshot();
  });

  test('Button gets disabled when passwords are changed and dont match', () => {
    render(<SignUpModal />);
    const textInputs = screen.queryAllByRole('textbox');
    const passwordInputs = screen.queryAllByLabelText(/password/i);

    // Fill in form
    userEvent.type(textInputs[0], 'mockUser');
    userEvent.type(textInputs[1], 'mock@mock.com');
    passwordInputs.forEach((input) => userEvent.type(input, 'password'));
    userEvent.type(passwordInputs[0], '{backspace}2');

    // Button should be enabled
    const button = screen.getByRole('button', { name: 'Sign Up' });
    expect(button).toHaveStyle('opacity: 0.5');
    expect(button).toHaveProperty('disabled', true);

    expect(container).toMatchSnapshot();
  });
});