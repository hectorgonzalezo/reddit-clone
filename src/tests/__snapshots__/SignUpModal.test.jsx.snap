// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Styled button works After filling form correctly, button gets enabled 1`] = `undefined`;

exports[`Styled button works Button doesnt get enabled if mail is missing @ 1`] = `undefined`;

exports[`Styled button works Button doesnt get enabled if passwords are different 1`] = `undefined`;

exports[`Styled button works Button gets disabled when passwords are changed and dont match 1`] = `undefined`;

exports[`Styled button works Renders with sign up button disabled 1`] = `undefined`;
